src/Build/BuildCheck/Infrastructure/CustomConfigurationData.cs

```diff
@@ -55,10 +65,40 @@ public override bool Equals(object? obj)
             return false;
         }
 
-        return Equals((CustomConfigurationData)obj);
-    }
+        var customConfigObj = (CustomConfigurationData) obj;
+
+        if(customConfigObj.RuleId != RuleId)
+        {
+            return false;
+        }
 
-    protected bool Equals(CustomConfigurationData other) => Equals(ConfigurationData, other.ConfigurationData);
+        // validate keys and values
+        if (customConfigObj.ConfigurationData != null && ConfigurationData != null)
+        {
+            if (!customConfigObj.ConfigurationData.Keys.SequenceEqual(ConfigurationData.Keys))
+            {
+                return false;
+            }
+
+            var keys = customConfigObj.ConfigurationData.Keys;
+            foreach (var key in keys)
+            {
+                if (customConfigObj.ConfigurationData[key] != ConfigurationData[key])
+                {
+                    return false;
+                }
+            }
+        }else if (customConfigObj.ConfigurationData == null && ConfigurationData == null)
```

Comment:
nit: Formatting
```suggestion
        }
        else if (customConfigObj.ConfigurationData == null && ConfigurationData == null)
```

Reply:
Another option:
```suggestion
        }
        else
        {
            return customConfigObj.ConfigurationData == null && ConfigurationData == null;
        }
```
to replace all the `return true` / `return false` below.

Reply:
Fixed the formatting . Thank you